specification/dsdl/architecture.tex

```diff
@@ -0,0 +1,233 @@
+
+Every data type definition is assigned a major and minor version number pair.
+In order to uniquely identify a data type definition, its version numbers must be specified.
+In the following text, the term \emph{version} without a majority qualifier refers to
+a pair of major and minor version numbers.
+
+Valid data type version numbers range from 0 to 255, inclusively.
+A data type version where both major and minor components are zero is not allowed.
+
+\subsection{File hierarchy}
+
+DSDL data type definitions are contained in UTF-8 encoded text files with a file name extension \verb|.uavcan|.
+
+One file defines exactly one version of a data type,
+meaning that each combination of major and minor version numbers must be unique per data type name.
+There may be an arbitrary number of versions of the same data type defined alongside each other,
+provided that each version is defined at most once.
+Version number sequences can be non-contiguous,
+meaning that it is allowed to skip version numbers or remove existing definitions that are neither oldest nor newest.
+
+A data type definition may have an optional fixed port ID\footnote{Chapter \ref{sec:basic_concepts}.} value specified.
+
+The name of a data type definition file is constructed from the following entities
+joined via the ASCII dot character ``\verb|.|'' (ASCII code 46), in the specified order:
+\begin{itemize}
+    \item Fixed port ID in decimal notation, unless a fixed port ID is not provided for this definition.
+    \item Short name of the data type (mandatory, always non-empty).
+    \item Major version number in decimal notation (mandatory).
+    \item Minor version number in decimal notation (mandatory).
+    \item File name extension ``\verb|uavcan|'' (mandatory).
+\end{itemize}
+
+\begin{figure}[H]
+    $$
+    \overbrace{%
+        \underbrace{\texttt{\huge{432}}}_{\substack{\text{fixed} \\ \text{port ID}}}%
+        \texttt{\huge{.}}%
+    }^{\text{optional}}%
+    \overbrace{%
+        \underbrace{\texttt{\huge{GetInfo}}}_{\substack{\text{short name}}}%
+        \texttt{\huge{.}}%
+        \underbrace{\texttt{\huge{1.0}}}_{\substack{\text{version} \\ \text{numbers}}}%
+        \texttt{\huge{.}}%
+        \underbrace{\texttt{\huge{uavcan}}}_{\text{file extension}}%
+    }^{\text{mandatory}}
+    $$
+    \caption{Data type definition file name structure.\label{fig:dsdl_definition_file_name_structure}}
+\end{figure}
+
+DSDL namespaces are represented as directories, where one directory defines exactly one namespace, possibly nested.
+The name of the directory defines the name of its data type name component.
+A directory defining a namespace will always define said namespace in its entirety,
+meaning that the contents of a namespace cannot be spread across different directories sharing the same name.
+One directory cannot define more than one level of
+nesting\footnote{For example, ``\texttt{foo.bar}'' is not a valid directory name.
```

Comment:
But most operating systems allow this as a folder name. Should we define a standard transformation from valid posix path names into valid dsdl that can be performed automatically as a pre-processing step (i.e. before we put the type into the grammar parser we perform a platform to internal representation transform). So "foo.bar" would become something like "foo_bar" etc.

Reply:
I am not sure what would we gain by doing so? Implicit transformations would make the relationship between DSDL names and generated type names harder to follow. The objective here is to say that if one wants to define a data type `vendor.foo.bar.Baz` so that the namespaces `foo` and `bar` are otherwise empty, they would have to create a dedicated directory `foo/` and then `foo/bar/` instead of just `foo.bar/`. The latter form could be supported but it greatly complicates the mapping between namespaces and directories. Additionally, it would be unclear how to represent that in the generated code: using identical shortcuts, or by unrolling the directory tree into one-level-per-directory form?

Reply:
I'm not saying we support a shortcut where `vendor/foo.bar` defines a <strike>`vendor.foo.bar`</strike> namespace. I'm suggesting we may want to consider the frustration of a user where they have `google.com/wing` and expect the namespaces to be `"google.com".wing`. We could decide to simply say all `.` in a folder name become underscores. So wing would generate `google_com.wing` namespaces. Of course this is asymmetric so I can be convinced that it's just simpler to throw a very nicely worded error about why the dsdl generator refused to continue.

Reply:
Is explicit not better than implicit?

We state that one can't use anything except `[a-zA-Z0-9_]` for a namespace name. If the user wants to name their namespace after some obscure Internet startup, they will have to replace their dotcoms with underscore-coms manually.

Reply:
okaaaaaaay